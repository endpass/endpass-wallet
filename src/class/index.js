export { providerFactory } from './provider';
export { default as Transaction } from './Transaction';
export {
  default as TransactionFactory,
} from './transaction/TransactionFactory';
export { default as EventEmitter } from './EventEmitter';
export { NotificationError } from './error';
export { default as Address } from './address';
export { default as Wallet } from './wallet';
export { ENSResolver } from './ens';
export { Token } from './Token';
export { default as ERC20Token } from './erc20';
export { default as TrezorWallet } from './hardwareWallet/TrezorWallet';
export { default as LedgerWallet } from './hardwareWallet/LedgerWallet';
