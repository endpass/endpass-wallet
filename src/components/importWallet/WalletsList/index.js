export { default } from './WalletsList.vue';
