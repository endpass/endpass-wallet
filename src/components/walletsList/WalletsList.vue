<template>
  <div class="column">
    <div
      v-if="!isLoading && !addresses.length"
      class="buttons"
    >
      <v-button
        :loading="isImporting"
        class-name="is-primary is-medium"
        @click="getNextAddressess"
      >
        Load wallets
      </v-button>
    </div>
    <v-spinner
      v-if="isLoading"
      :is-loading="true"
      label="Please, allow access for import"
      is-label-under-spinner
    />
    <div v-show="!isLoading && addresses.length">
      <wallet-item
        v-for="address in addresses"
        :key="address"
        :address="address"
        :is-active="address === activeAddress"
        @click="setActiveAddress(address)"
      />
      <v-pagination @input="changePage" />
      <div class="buttons">
        <v-button
          :disabled="!activeAddress"
          :loading="isImporting"
          class-name="is-primary is-medium"
          data-test="import-wallet-button"
          @click="handleImport"
        >
          Import
        </v-button>
      </div>
    </div>
    <password-modal
      v-if="isPasswordModal"
      @close="togglePasswordModal"
      @confirm="handleImportByPassword"
    />
  </div>
</template>

<script>
import { mapActions, mapGetters } from 'vuex';
import modalMixin from '@/mixins/modal';
import PasswordModal from '@/components/modal/PasswordModal';
import WalletItem from './WalletItem.vue';

export default {
  name: 'WalletsList',
  props: {
    type: {
      type: String,
      required: true,
    },
    autoLoad: {
      type: Boolean,
      default: false,
    },
  },
  data: () => ({
    activeAddress: null,
    isLoading: false,
    isImporting: false,
    isPasswordModal: false,
    addresses: [],
    offset: 0,
    limit: 10,
  }),
  computed: {
    ...mapGetters('accounts', ['isHDv3WalletByType']),
  },
  methods: {
    ...mapActions('accounts', [
      'addPublicWallet',
      'getNextWalletsFromHd',
      'addHdChildWallets',
    ]),

    async getNextAddressess({ offset = this.offset, limit = this.limit } = {}) {
      this.isLoading = true;

      try {
        this.addresses = await this.getNextWalletsFromHd({
          offset,
          limit,
          walletType: this.type,
        });
      } catch (error) {
        this.$notify(error);
      } finally {
        this.isLoading = false;
      }
    },
    async changePage(page) {
      this.offset = this.limit * (page - 1);
      await this.getNextAddressess();
    },
    async handleImportByPassword(password) {
      await this.addWallet(password);
    },
    async handleImport() {
      if (this.isHDv3WalletByType(this.type)) {
        this.isPasswordModal = true;
      } else {
        await this.addWallet();
      }
    },
    async addWallet(password) {
      this.isImporting = true;
      try {
        const { activeAddress: address, type } = this;
        const index = this.addresses.indexOf(address);
        if (index === -1) {
          throw new Error('Address not selected');
        }

        if (this.isHDv3WalletByType(type)) {
          await this.addHdChildWallets({
            address,
            index,
            password,
            type,
          });
        } else {
          const info = { type, index };
          await this.addPublicWallet({ address, info });
        }
        this.$router.push('/');
      } catch (e) {
        this.$notify({
          title: 'Importing error',
          text: 'An error occurred while importing wallet. Please try again.',
          type: 'is-danger',
        });
      } finally {
        this.isImporting = false;
      }
    },
    setActiveAddress(address) {
      this.activeAddress = address;
      this.$emit('select', address);
    },
  },
  mounted() {
    if (this.autoLoad) {
      this.getNextAddressess();
    }
  },
  mixins: [modalMixin],
  components: {
    WalletItem,
    PasswordModal,
  },
};
</script>

<style lang="scss"></style>
