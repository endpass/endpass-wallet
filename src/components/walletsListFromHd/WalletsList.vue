<template>
  <div class="column">
    <div
      v-if="!isGettingAddresses && isEmptyAddresses"
      class="buttons"
    >
      <v-button
        :loading="isImporting"
        class-name="is-primary is-medium"
        @click="getNextAddressess"
      >
        Load wallets
      </v-button>
    </div>
    <v-spinner
      v-if="isGettingAddresses"
      :is-loading="true"
      label="Please, allow access for import"
      is-label-under-spinner
    />
    <div v-show="!isGettingAddresses && !isEmptyAddresses">
      <wallet-item
        v-for="(address, index) in addresses"
        :key="address"
        :address="address"
        :is-used="isUsedWallet(address)"
        :is-active="address === selectedAddress"
        @click="updateSelectedAddress(address, index)"
      />
      <v-pagination @input="changePage" />
      <slot name="buttons" />
    </div>
  </div>
</template>

<script>
import { mapActions, mapState } from 'vuex';
import WalletItem from './WalletItem.vue';

export default {
  name: 'WalletsList',
  props: {
    type: {
      type: String,
      required: true,
    },
    autoLoad: {
      type: Boolean,
      default: false,
    },
    isImporting: {
      type: Boolean,
      default: false,
    },
    isGettingAddresses: {
      type: Boolean,
      default: false,
    },
  },
  data: () => ({
    offset: 0,
    limit: 10,
    addresses: [],
    selectedAddress: null,
  }),
  computed: {
    isEmptyAddresses() {
      return this.addresses.length === 0;
    },
    ...mapState('accounts', ['wallets']),
  },
  methods: {
    ...mapActions('accounts', ['getNextWalletsFromHd']),

    async getNextAddressess({ offset = this.offset, limit = this.limit } = {}) {
      this.$emit('input', true);
      const { type: walletType } = this;
      try {
        this.addresses = await this.getNextWalletsFromHd({
          offset,
          limit,
          walletType,
        });
      } catch (error) {
        this.$notify(error);
      } finally {
        this.$emit('input', false);
      }
    },

    async changePage(page) {
      this.offset = this.limit * (page - 1);
      await this.getNextAddressess();
    },

    /**
     * @param {string} address
     * @param {number} index
     */
    updateSelectedAddress(address, index) {
      this.selectedAddress = address;
      this.$emit('select', { address, index: index + this.offset });
    },

    /**
     * @param {string} address
     * @returns {boolean}
     */
    isUsedWallet(address) {
      const account = this.wallets[address];
      return account ? !account.isPublic && !account.isHardware : false;
    },
  },
  mounted() {
    if (this.autoLoad) {
      this.getNextAddressess();
    }
  },
  components: { WalletItem },
  model: {
    prop: 'isGettingAddresses',
    event: 'input',
  },
};
</script>
