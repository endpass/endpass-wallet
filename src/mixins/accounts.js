// Mixin for common methods for Ethereum accounts
import web3 from 'web3';

const { fromWei } = web3.utils;

export default {
  computed: {
    address() {
      return this.$store.state.accounts.address;
    },

    balance() {
      return this.$store.state.accounts.balance === null
        ? null
        : fromWei(this.$store.getters['accounts/balance']);
    },
  },
};
