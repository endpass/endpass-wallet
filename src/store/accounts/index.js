import actions from './actions';
import getters from './getters';
import mutations from './mutations';

/**
 * @namespace
 * @property {Boolean} isInited Status of initializing accounts module
 * @property {Object} hdKey The encrypted xprv key for the user's hd wallet as a V3 keystore object.
 *  All accounts are direct child accounts.
 * @property {Object} hadrwareXpub The user's hardware xpub, keyed by wallet type
 * @property {Object} wallets The user's wallets, mapped by address
 * @property {String} address Current wallet address
 * @property {String} balance Currect account balance
 */
const state = {
  isInited: false,
  hardwareXpub: {},
  wallets: {},
  hdKey: null,
  address: null,
  balance: null,
};

export default {
  namespaced: true,
  state,
  actions,
  getters,
  mutations,
};
