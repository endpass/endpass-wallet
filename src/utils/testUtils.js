import { kebabCase } from 'lodash';
import { getFrom } from '@/utils/objects';

function convertDirectivesToAttrs(directives) {
  const attrs = {};

  if (directives) {
    directives.forEach(({ rawName, expression }) => {
      attrs[rawName] = expression;
    });
  }

  return attrs;
}

function convertListenersToAttrs(listeners) {
  const attrs = {};

  if (listeners) {
    Object.keys(listeners).forEach(key => {
      const handlers = [].concat(listeners[key]);
      const handler = handlers.find(h => {
        const name = getFrom(h, 'fns.name', 'name') || '';

        return name.includes('bound ');
      });

      if (handler) {
        attrs[`v-on:${key}`] = (handler.fns || handler).name;
      }
    });
  }

  return attrs;
}

function convertModelToAttrs(model) {
  return model ? { 'v-model': model.expression } : {};
}

function convertSlotToAttrs(slot) {
  return slot ? { slot } : {};
}

function generateElement(createElement, vnode) {
  return vnode.componentOptions
    ? generateElementFromComponent(createElement, vnode)
    : generateElementFromHTML(createElement, vnode);
}

function getChildrenElements(createElement, children) {
  return children
    ? children.map(child => generateElement(createElement, child))
    : undefined;
}

function generateElementFromComponent(
  createElement,
  { data, componentOptions },
) {
  const { attrs, directives, model } = data;
  const { tag, listeners, children } = componentOptions;
  const componentData = {
    ...data,
    attrs: {
      ...attrs,
      ...convertDirectivesToAttrs(directives),
      ...convertListenersToAttrs(listeners),
      ...convertModelToAttrs(model),
      class: data.staticClass,
    },
    hook: undefined,
    staticClass: undefined,
  };

  if (listeners) {
    componentData.on = Object.assign({}, listeners);
  }

  return createElement(
    tag,
    componentData,
    getChildrenElements(createElement, children),
  );
}

function generateElementFromHTML(createElement, vnode) {
  if (!vnode.data) {
    return vnode;
  }

  const { tag, data, children } = vnode;
  const { directives, on, model, attrs = {}, slot } = data;

  return createElement(
    tag,
    {
      ...data,
      attrs: {
        ...attrs,
        ...convertDirectivesToAttrs(directives),
        ...convertListenersToAttrs(on),
        ...convertModelToAttrs(model),
        ...convertSlotToAttrs(slot),
      },
    },
    getChildrenElements(createElement, children),
  );
}

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let elementName;

    if (stubComponent.name) {
      elementName = kebabCase(stubComponent.name);
    }

    return Object.assign(stubs, {
      [elementName]: {
        render(createElement) {
          return generateElementFromComponent(createElement, this.$vnode);
        },
      },
    });
  }, {});
}

export default {
  generateStubs,
};
