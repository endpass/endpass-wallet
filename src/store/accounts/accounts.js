import { userService } from '@/services';
import storage from '@/services/storage';
import web3 from 'web3';
import Bip39 from 'bip39';
import HDKey from 'ethereumjs-wallet/hdkey';
import { hdKeyMnemonic, kdfParams } from '@/config';
import EthWallet from 'ethereumjs-wallet';
import { SAVE_TOKENS } from '@/store/tokens/mutations-types';
import { Wallet, Address } from '@/class';
import { BigNumber } from 'bignumber.js';
import keystore from '@/utils/keystore';

export default {
  namespaced: true,
  state: {
    // TODO move to user store?
    email: null,
    // The encrypted xprv key for the user's hd wallet as a V3 keystore
    // object. All accounts are direct child accounts.
    hdKey: null,
    // The user's wallets, keyed by address
    wallets: {},
    // The currently selected wallet
    wallet: null,
    address: null,
    balance: null,
    price: null,
    // prettier-ignore
    availableCurrencies: ['USD', 'AUD', 'BRL', 'CAD', 'CHF', 'CLP', 'CNY', 'CZK', 'DKK', 'EUR', 'GBP', 'HKD', 'HUF', 'IDR', 'ILS', 'INR', 'JPY', 'KRW', 'MXN', 'MYR', 'NOK', 'NZD', 'PHP', 'PKR', 'PLN', 'RUB', 'SEK', 'SGD', 'THB', 'TRY', 'TWD', 'ZAR'],
    settings: {
      fiatCurrency: 'USD',
    },
    otpSettings: {
      secret: null,
      status: null,
    },
  },
  getters: {
    isPublicAccount(state) {
      return (
        state.address instanceof Address && !(state.wallet instanceof Wallet)
      );
    },
    // TODO move to user store?
    isLoggedIn(state) {
      return !!state.email;
    },
    balance(state, getters, rootState, rootGetters) {
      if (state.balance === null) return null;

      const pendingBalance = rootGetters['transactions/pendingBalance'];
      const balanceWei = BigNumber(state.balance || '0')
        .minus(pendingBalance)
        .toFixed();

      return web3.utils.fromWei(balanceWei);
    },
    // Returns a decrypted HD Wallet
    hdWallet: state => password => {
      if (!state.hdKey) {
        return null;
      }
      return keystore.decryptHDWallet(password, state.hdKey);
    },
  },
  mutations: {
    setAddress(state, addressString) {
      state.address = new Address(addressString);
    },
    setWallet(state, wallet) {
      state.wallet = wallet;
    },
    addWallet(state, walletV3) {
      const wallet = new Wallet(walletV3);
      const { address } = walletV3;
      state.wallets = {
        ...state.wallets,
        [address]: wallet,
      };
    },
    // Saves the encrypted HD wallet key in V3 keystore format
    // Formerly addHdWallet
    setHdKey(state, key) {
      state.hdKey = key;
    },
    setBalance(state, balance) {
      state.balance = balance;
    },
    setPrice(state, price) {
      state.price = price;
    },
    setSettings(state, settings) {
      state.settings = JSON.parse(JSON.stringify(settings));
    },
    setEmail(state, email) {
      state.email = email;
    },
    setOtpSettings(state, otpSettings) {
      state.otpSettings = otpSettings;
    },
  },
  actions: {
    selectWallet({ commit, state, dispatch }, address) {
      commit('setWallet', state.wallets[address]);
      commit('setAddress', address);
      return dispatch(
        'tokens/subscribeOnTokensBalancesUpdates',
        {},
        { root: true },
      );
    },
    addWallet({ commit, dispatch, state }, json) {
      json.address = web3.utils.toChecksumAddress(json.address);
      commit('addWallet', json);

      return userService
        .setAccount(json.address, json)
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    addWalletAndSelect({ dispatch }, json) {
      return dispatch('addWallet', json)
        .then(() =>
          dispatch('selectWallet', web3.utils.toChecksumAddress(json.address)),
        )
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    // Import wallet from json V3 keystore
    async addWalletWithV3({ commit, dispatch }, { json, password }) {
      try {
        const wallet = new Wallet(json);
        let privateKey = await wallet.getPrivateKey(password);
        return dispatch('addWalletWithPrivateKey', privateKey);
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
    async addWalletWithPrivateKey(
      { commit, dispatch },
      { privateKey, password },
    ) {
      try {
        const wallet = EthWallet.fromPrivateKey(Buffer.from(privateKey, 'hex'));
        const json = keystore.encryptWallet(password, wallet);
        return dispatch('addWalletAndSelect', json);
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
    generateWallet({ commit, dispatch, state, getters }, password) {
      if (!state.hdKey) {
        return;
      }
      try {
        let hdWallet = getters.hdWallet(password);
        let i = Object.keys(state.wallets).length;
        let wallet = hdWallet.deriveChild(i).getWallet();
        let json = keystore.encryptWallet(password, wallet);
        return dispatch('addWalletAndSelect', json);
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
    // Saves HD wallet's extended keys on the server
    saveHdWallet({ commit, dispatch, state }, json) {
      return userService.setAccount(json.address, json);
    },
    addHdWallet({ commit, dispatch }, { key, password }) {
      const seed = Bip39.mnemonicToSeed(key);
      const hdKey = HDKey.fromMasterSeed(seed);
      const hdWallet = hdKey.derivePath(hdKeyMnemonic.path);
      // Encrypt extended private key
      const json = keystore.encryptHDWallet(password, hdWallet);
      commit('setHdKey', json);

      // Save HD keys and generate the first child wallet
      return dispatch('saveHdWallet', json)
        .then(() => dispatch('generateWallet', password))
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    async addMultiHdWallet({ commit, dispatch, rootState }, { key, password }) {
      const seed = Bip39.mnemonicToSeed(key);
      const hdKey = HDKey.fromMasterSeed(seed);
      const hdWallet = hdKey.derivePath(hdKeyMnemonic.path);

      /* eslint-disable no-await-in-loop */
      for (let index = 0; index < 5; index++) {
        const wallet = hdWallet.deriveChild(index).getWallet();
        const walletV3 = wallet.toV3(Buffer.from(password), kdfParams);
        const { address } = walletV3;

        if (index !== 0) {
          dispatch('addWallet', walletV3);
        } else {
          dispatch('addWalletAndSelect', walletV3);
        }

        try {
          const balance = await rootState.web3.web3.eth.getBalance(address);

          if (balance === '0') {
            break;
          }
        } catch (e) {
          break;
        }
      }
      /* eslint-enable no-await-in-loop */
    },
    updateBalance({ commit, dispatch, state, rootState }) {
      if (state.address) {
        const address = state.address.getChecksumAddressString();

        return rootState.web3.web3.eth
          .getBalance(address)
          .then(balance => commit('setBalance', balance))
          .catch(e => dispatch('errors/emitError', e, { root: true }));
      }
    },
    updateSettings({ commit, dispatch }, settings) {
      commit('setSettings', settings);
      return storage
        .write('settings', settings)
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    validatePassword({ state }, password) {
      return state.wallet.validatePassword(password);
    },
    login({ commit, dispatch }, email) {
      return userService.login(email);
    },
    logout({ commit, dispatch }) {
      commit('setEmail', null);
      return Promise.all([storage.clear(), userService.logout()])
        .then(() => window.location.reload())
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    loginViaOTP({}, { code, email }) {
      return userService.loginViaOTP(code, email);
    },
    getOtpSettings({ commit, dispatch }) {
      return userService
        .getOtpSettings()
        .then(otpSettings => commit('setOtpSettings', otpSettings))
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    setOtpSettings({ commit, dispatch }, { secret, code }) {
      return userService
        .setOtpSettings(secret, code)
        .then(() => commit('setOtpSettings', { status: 'enabled' }))
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    deleteOtpSettings({ commit, dispatch }, { code }) {
      return userService
        .deleteOtpSettings(code)
        .then(() => {
          commit('setOtpSettings', {});
          dispatch('getOtpSettings');
        })
        .catch(e => dispatch('errors/emitError', e, { root: true }));
    },
    async init({ commit, dispatch }) {
      try {
        let [settings, email] = await Promise.all([
          userService.getSettings(),
          storage.read('email'),
        ]);

        commit(`tokens/${SAVE_TOKENS}`, settings.tokens || {}, { root: true });
        commit('setEmail', email);

        if (settings) {
          commit('setSettings', settings);
        }

        if (!email) {
          storage.disableRemote();
          return null;
        }

        // Fetch and save HD wallet
        let hdKey = await userService.getHDKey();
        if (hdKey) {
          commit('setHdKey', hdKey);
        }

        // Fetch and save regular accounts
        let accounts = await userService.getV3Accounts();
        if (accounts && accounts.length) {
          accounts.forEach(wallet => commit('addWallet', wallet));
          await dispatch('selectWallet', accounts[0].address);
        }
      } catch (e) {
        return dispatch('errors/emitError', e, { root: true });
      }
    },
  },
};
