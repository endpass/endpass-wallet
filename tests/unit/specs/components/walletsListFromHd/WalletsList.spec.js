import { createLocalVue } from '@vue/test-utils';
import Notifications from 'vue-notification';
import Vuex from 'vuex';
import VeeValidate from 'vee-validate';
import UIComponents from '@endpass/ui';
import { wrapShallowMountFactory } from '@/testUtils';

import WalletsList from '@/components/walletsListFromHd/WalletsList';
import { Wallet } from '@/class';
import { v3 } from 'fixtures/accounts';

const WALLET_TYPES = Wallet.getTypes();

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VeeValidate);
localVue.use(Notifications);
localVue.use(UIComponents);

jest.useFakeTimers();

describe('WalletsList', () => {
  let wrapper;
  let wallets;
  let wrapperFactory;
  let actions;

  beforeEach(() => {
    actions = {
      addPublicWallet: jest.fn(),
      getNextWalletsFromHd: jest.fn(),
      addHdChildWallets: jest.fn(),
    };
    const storeOptions = {
      modules: {
        accounts: {
          namespaced: true,
          actions,
          getters: {
            isHDv3WalletByType: jest.fn(),
          },
        },
      },
    };
    const store = new Vuex.Store(storeOptions);
    wallets = {
      [v3.address]: {
        ...v3,
        isPublic: false,
        isHardware: false,
      },
    };
    wrapperFactory = wrapShallowMountFactory(WalletsList, {
      localVue,
      store,
      // sync: false,
      propsData: {
        type: WALLET_TYPES.HD_PUBLIC,
      },
      computed: {
        wallets,
      },
    });
    wrapper = wrapperFactory();
  });

  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('WalletsList');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render initial state of the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('methods', () => {
    const emptyArray = [];

    describe('getNextAddressess', () => {
      it('should call vuex getNextWalletsFromHd with correct arguments', async () => {
        expect.assertions(3);

        wrapper.setMethods({
          getNextWalletsFromHd: jest.fn().mockResolvedValue([]),
        });

        wrapper.find('v-pagination-stub').vm.$emit('offset', 0);
        await global.flushPromises();

        expect(wrapper.vm.addresses).toEqual(emptyArray);
        expect(wrapper.vm.getNextWalletsFromHd).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.getNextWalletsFromHd).toBeCalledWith({
          offset: 0,
          limit: wrapper.vm.limit,
          walletType: WALLET_TYPES.HD_PUBLIC,
        });
      });

      it('should call changePage with correct arguments', async () => {
        expect.assertions(3);

        wrapper.setMethods({
          getNextWalletsFromHd: jest.fn().mockResolvedValue([]),
        });

        wrapper.find('v-pagination-stub').vm.$emit('offset', 20);
        await global.flushPromises();
        expect(wrapper.vm.addresses).toEqual(emptyArray);
        expect(wrapper.vm.getNextWalletsFromHd).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.getNextWalletsFromHd).toBeCalledWith({
          offset: 20,
          limit: 10,
          walletType: WALLET_TYPES.HD_PUBLIC,
        });
      });
    });

    describe('updateSelectedAddress', () => {
      it('should emit correct address and index', () => {
        const addresses = ['a', 'b', 'c'];
        wrapper.setData({
          addresses,
        });

        addresses.forEach((address, index) => {
          wrapper.setData({
            offset: 10 * index,
          });
          wrapper
            .findAll('wallet-item-stub')
            .at(index)
            .vm.$emit('click');
          expect(wrapper.vm.selectedAddress).toBe(address);
          const payload = { address, index: index + wrapper.vm.offset };
          expect(wrapper.emitted('select')[index]).toEqual([payload]);
        });
      });
    });

    describe('usedWallet', () => {
      it('should check if wallet already added', () => {
        expect(wrapper.vm.isUsedWallet('foo')).toBeFalsy();
        expect(wrapper.vm.isUsedWallet(v3.address)).toBeTruthy();
        wallets[v3.address].isPublic = true;
        expect(wrapper.vm.isUsedWallet(v3.address)).toBeFalsy();
        wallets[v3.address].isPublic = false;
        wallets[v3.address].isHardware = true;
        expect(wrapper.vm.isUsedWallet(v3.address)).toBeFalsy();
        wallets[v3.address].isPublic = true;
        expect(wrapper.vm.isUsedWallet(v3.address)).toBeFalsy();
      });
    });
  });
});
