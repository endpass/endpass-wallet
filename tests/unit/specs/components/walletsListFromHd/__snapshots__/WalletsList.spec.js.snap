// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletsList render should render initial state of the component 1`] = `
<div
  class="column"
>
  <div
    class="buttons"
  >
    <v-button-stub
      classname="is-primary is-medium"
    >
      
      Load wallets
    
    </v-button-stub>
  </div>
   
  <!---->
   
  <div
    style="display: none;"
  >
     
    <v-pagination-stub
      limit="10"
      offset="0"
      pagesvisible="0"
      total="0"
    />
     
  </div>
</div>
`;
